OCPBUGS-57348: add MCO operator manifest for boot image management

patrickdillon opened the pull request:
Adds manifest generation for MCO configuration. Currently the manifest is only generated when compute node custom boot images are specified, in order to disable MCO management of those boot images.

For example, aws install config with compute ami set:

```yaml
compute:
- architecture: amd64
  hyperthreading: Enabled
  name: worker
  platform:
    aws:
      amiID: ami-08f1807771f4e468b
  replicas: 3
```
 
Produces the following manifest

```shell
cat test/manifests/cluster-mco-02-config.yml
```

```yaml
apiVersion: operator.openshift.io/v1
kind: MachineConfiguration
metadata:
  name: cluster
spec:
  logLevel: Normal
  operatorLogLevel: Normal
  managedBootImages:
    machineManagers:
      - resource: machinesets
        apiGroup: machine.openshift.io
        selection:
          mode: None
```

djoshy commented on pkg/asset/manifests/mco.go:
```diff
@@ -0,0 +1,188 @@
+		return true, true
+	}
+
+	if cp := ic.ControlPlane; cp != nil && cp.Platform.GCP != nil && cp.Platform.GCP.OSImage != nil {
```
The control plane machinesets are a different resource: `ControlPlaneMachineSets`, and the MCO does not currently manage them(https://issues.redhat.com/browse/MCO-1007). So we could simplify this bit and only worry about compute.

patrickdillon replied:
Working on it now, and I am able to make it simpler. Thanks!

djoshy commented on pkg/asset/manifests/mco.go:
```diff
@@ -0,0 +1,188 @@
+}
+
+func generateMCOCfg(installConfig *types.InstallConfig) *operatorv1.MachineConfiguration {
+	cfg := &operatorv1.MachineConfiguration{
```
It would be useful if we can generate an annotation or a marker of some sort to indicate that the installer did this and not the user.

patrickdillon replied:
Closing this, based on trevor's comment

wking commented on pkg/asset/manifests/mco.go:
```diff
@@ -0,0 +1,172 @@
+)
+
+const (
+	roleLabel = "machine.openshift.io/cluster-api-machine-role"
```
[lint][1]:

```
pkg/asset/manifests/mco.go:24:2: const `roleLabel` is unused (unused)
	roleLabel = "machine.openshift.io/cluster-api-machine-role"
	^
```

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9783/pull-ci-openshift-installer-main-golint/1933255862084702208#1:build-log.txt%3A33-35

wking commented on pkg/asset/manifests/mco.go:
```diff
@@ -0,0 +1,172 @@
+	}
+
+	if customCPImg {
+		// TODO: when CPMS boot image management is implemented in MCO-1007.
```
[lint][1]:

```console
pkg/asset/manifests/mco.go:157:2: SA9003: empty branch (staticcheck)
	if customCPImg {
	^
```

Personally, I'd drop this `if`, shift the comment to the commit message, and add a note to the MCO-1007 Jira ticket to ask the MCO folks to reach out to the installer if/when they get around to ControlPlaneMachineSet boot-image management.

[1]: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_installer/9783/pull-ci-openshift-installer-main-golint/1933255862084702208#1:build-log.txt%3A36-38

patrickdillon commented:
Ok I have reworked the PR, because cluster-bootstrap was not able to successfully apply the manifest created in the original implementation. I confirmed in testing what @djoshy suspected: when golang serializes the empty values of the mco configuration, it leaves a lot of empty values, and API server validation fails on them.

So in order to create the minimally specified manifest (in the updated description above), I switched the implementation to use a go template file. I have added this as a fixup commit ATM and will squash it later, once we align on this approach.

patrickdillon commented:
I have the lint fixes committed locally but will wait to push them until I get more feedback. This is a substantial change from the first iteration, so I wanted to get it up for review.  I also still need to rework the mco tests; which I think I can mostly reuse.

Local testing shows the configuration I would expect in the cluster:

```
% oc describe machineconfiguration cluster 
Name:         cluster
Namespace:    
Labels:       <none>
Annotations:  <none>
API Version:  operator.openshift.io/v1
Kind:         MachineConfiguration
Metadata:
  Creation Timestamp:  2025-06-15T23:39:28Z
  Generation:          1
  Resource Version:    2513
  UID:                 40bfcf3b-fd93-4066-9006-f9e37cc0808e
Spec:
  Log Level:  Normal
  Managed Boot Images:
    Machine Managers:
      API Group:  machine.openshift.io
      Resource:   machinesets
      Selection:
        Mode:          None
  Operator Log Level:  Normal
Events:                <none>
```

wking commented on pkg/asset/templates/content/manifests/cluster-mco-config.go:
```diff
@@ -0,0 +1,65 @@
+)
+
+const (
+	mcoConfigTemplateFileName = "cluster-mco-02-config.yaml.template"
```
Where does the eventual generated manifest's filename sort vs. [the installer-generated MachineSet manifests][1]?  Do we want a `90_` prefix or something to ensure we sort the MachineConfiguration first, so the MCO knows to leave MachineSets alone before the MachineSets are created in the cluster?

[1]: https://github.com/openshift/installer/blob/55b35185c79710706be350d384fc35c2afb322c3/pkg/asset/machines/worker.go#L71-L72

djoshy commented:
/lgtm

The MCO manifests and tests seems sane to me. I'm not as familiar with the manifest generation flow, so I will defer that to the installer folks. Thanks for putting this together!

patrickdillon commented:
/retest-required

patrickdillon commented:
#9793 has merged for unit tests

/retest-required

patrickdillon commented:
/label tide/merge-method-squash
for fixups

patrickdillon commented:
@djoshy would you mind bumping lgtm?

patrickdillon commented:
/cherry-pick release-4.19

djoshy commented:
/lgtm

patrickdillon commented:
/approve

gpei commented:
/label qe-approved
